x/bank/module.go

```diff
@@ -100,6 +101,7 @@ type AppModule struct {
 func (am AppModule) RegisterServices(cfg module.Configurator) {
 	types.RegisterMsgServer(cfg.MsgServer(), keeper.NewMsgServerImpl(am.keeper))
 	types.RegisterQueryServer(cfg.QueryServer(), am.keeper)
+	cfg.RegisterMigration(types.ModuleName, 0, v042.MigrateStore)
```

Comment:
The store key and codec are in the keeper already so they can be passed from the keeper to the migration handler closure not from the module manager into the handler.

Reply:
And even if the keeper didn't have this stuff already we can add it to `Configurator` so that all services have it rather than separately to each migration handler. Does that make sense? I think it will be much cleaner this way. Otherwise every time we need some new bit of configuration stuff then we keep adding more params to migration handler. Instead we can keep the migration handler params minimal and pass around other static configuration higher up.

Reply:
> The store key and codec are in the keeper already so they can be passed from the keeper to the migration handler closure not from the module manager into the handler.

Re store key: it is in the keeper yes, but not accessible by the `AppModule` at this line (cannot do `am.keeper.storeKey`).

Happy to discuss it today (or feel free to push to this PR if it's faster).

Reply:
Yeah, so I think the right way is to make the migration a keeper method. 

Basically:
```go
cfg.RegisterMigration(types.ModuleName, 0, func(ctx sdk.Context) error {
  return am.keeper.Migrate0(ctx)
})
```

Reply:
Ah, ok, yeah that works, but it also pollutes the keeper. In bank, the Keeper is an interface, so we'd need to put it as a public method there, and it'd also be exposed to other modules.

I also like the fact that migration scripts are in each module's `legacy` folder.

I prefer the current approach, but if you think the keeper's way is cleaner, I'm okay to change it. (cc @robert-zaremba, maybe you have an opinion, since you reviewed this PR?)

Reply:
In ADR 033, we include both `StoreKey` and `Marshaler` on `Configurator` and we could explore that step. An alternative to polluting the keeper is just to put the storeKey and codec in `AppModule` when it gets created. Or we can just make a `MigrateKeeper` interface and cast to that and we still keep the scripts in legacy but import them from the keeper... But I really am pretty strongly opposed to making these parameters on the migration handler, it's just not the right place.

Reply:
> Or we can just make a MigrationKeeper interface and cast Keeper to that and we still keep the scripts in legacy but import them from the keeper

I went with this in 41e0339. I feel strongly about having those scripts in legacy, because I am copy/pasting legacy code into those folders.

Reply:
Yeah, I agree they should be in legacy.

What do you think about having StoreKey and Marshaler on Configurator?

One big thing I'm trying to pay attention to with APIs is not trying to pass too many things with function params. What ends up happening is that people say, oh we need just one more parameter and then the function signature changes. It's much less breaking to add fields to structs and methods to interfaces.

Reply:
This is worth keeping in mind: https://blog.golang.org/module-compatibility

Reply:
> What do you think about having StoreKey and Marshaler on Configurator?

Having `Marshaler` on Configurator is useless imo (for migrations), because most (all?) AppModuleBasics already have a `cdc`.

Would StoreKey on Configurator look like `func StoreKey(moduleName string) sdk.StoreKey`? How do you avoid moduleA doing in its AppModule: `b:=configurator.StoreKey("moduleB")`?